### ISSUE TITLE
Kubernetes V2 "Run Job" Stage

### ISSUE THREAD

**Issue by lwander** [spinnaker/spinnaker#2587]
It's not immediately clear how much this should mirror the V1 Run Job stage's functionality, mainly because Deploy (Manifest) allows you to do (mostly) the same set of things by just deploying a Pod.

What's missing from Deploy (Manifest):

1. Better insight into the state of your "job" while the stage is running.
2. Logs.
3. A nice UI.

IMO, we should fix 1, provide enough context to create custom log links for 2, and wait on 3 until we have clearer UI requirements.

**Comment by wjoel** [comment]
Access to logs and events during deployment would be useful for many kinds of resources.

Kubernetes jobs stay around after completion and keep the logs available (which is also why they need to be explicitly removed after completion), so that should be possible. I haven't actually checked if they show up in the UI, probably not?

**Comment by benjaminws** [comment]
Given our primary use case for this is to render our own manifests using helm, we need to be able to pass arguments to a pipeline stage like a chart version and output path for rendered manifests.

Additionally, you can see these two issues for other functionality we would like to see. #2552 and #2553

**Comment by wjoel** [comment]
Init containers and node selectors would probably be ok just with the manifests in v2.

Passing arguments brings us back to whether or not we should support something like SpEL or Jinja in manifest templates. Maybe if we allow it in "text" sources in the stage itself? That would be similar to other string fields in the Spinnaker UI (and in the pipeline template definitions) that allow SpEL/Jinja.

**Comment by wjoel** [comment]
I somehow missed https://github.com/spinnaker/orca/pull/1888, pipeline expressions in manifests are already evaluated.

**Comment by Rigdon** [comment]
Thanks @wjoel that may actually work out well enough for us with additional logging and insight into the deployed `Job` manifest.

**Comment by lwander** [comment]
I'll work on surfacing events in that details panel for the pipeline execution in service of this.

Can people interested chime-in on what their logging solution is, and what a query (deep link) into the system looks like? If it seems feasible, we can construct a set of canonical queries using SpEL to embed in the "comments" field to make looking up logs easy.

**Comment by ethanfrogers** [comment]
We're using Fluentd to forward logs into Elasticsearch, typically viewed via Kibana. When I go to look at logs I typically jump into Kibana and filter by `kubernetes_container_name` (Fluentd staples this to the log) or by some label value. Before V1 Run Job logs, I setup our logging annotation on the Run Job Pod and added a `app: migration` label to the Pod.

**Comment by lwander** [comment]
Nice, do you have a sample link?

**Comment by LouisStAmour** [comment]
Thought I'd raise it here, but I'm looking to replace Docker Hub or the legacy "Bake" step with lightweight Kubernetes jobs, an example could look like https://github.com/chgeuer/k8s_elixir/blob/master/build-job-parametrized.yaml

As to deleting jobs, it sounds like the best approach is to do so on a scheduled basis: https://github.com/hjacobs/kube-job-cleaner though there are real-time examples also: https://github.com/lwolf/kube-cleanup-operator

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 110 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by benjaminws** [comment]
At Datadog we have a pretty good workflow using kubernetes v2 provider + the kubernetes `Job` kind now. I'm not so sure this is needed, except for maybe historical reference?

**Comment by musabmasood** [comment]
I am using the latest Kubernetes stable helm chart. I don't see any "run job" stage in there. Has this stage been depreciated? Is Jenkins stage or Script stage the only option I have to run arbitrary scripts?

**Comment by lwander** [comment]
What Benjamin is saying is that he is using the "Deploy Manifest" stage
with a `Job` kind -- the stage will only complete if the job succeeds, and
fail if it times out, as well as surface links to logs for the pods running
the job.

**Comment by dchw** [comment]
To add, I have had great success using `library/bash` and passing the whole script as an argument. Ends up looking like most scripts you find in dockerfile. It nice because I can just write a script right there in the "Deploy Manifest" configuration with basically no effort.

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by ethanfrogers** [comment]
Been exploring this lately and wanted to raise a couple of questions/comments here - 
1) If we bring the Run Job stage back to V2 we have the opportunity to utilize Preconfigured Job stages (https://github.com/spinnaker/orca/pull/2540). This would allow users to have "native" stages that run as containers. 
2) The Run Job experience feels more integrated than deploying raw manifests. While you can deploy manifests to accomplish the same task I've found that it isn't quite what users are looking for when they want to just run a job.
3) If we were to bring this back would you want the ability to specify a `Job` spec  or a `Pod` spec? Perhaps both?

**Comment by avram** [comment]
I agree with 1 & 2 --- this should make it easier to have things run against K8s under the hood without the user having to think so much in terms of k8s YAML.

I would prefer to see this as a `Job` -- I've rarely run a `Pod` that I didn't regret.

**Comment by ethanfrogers** [comment]
Both can be useful, IMO. A `Job` is good if you want to ensure that your task finishes successfully and is restarted if not. I think a `Pod` aligns more closely to what one would expect when running a task in a pipeline - "do this thing and fail the pipeline if it fails". I don't think there's a way to simulate that using a `Job` (run _only_ once).

I'll see if I can make it work with both resources, however.

**Comment by dchw** [comment]
While we are talking about things that would be useful here:

- I would like to be able to see the logs (success or failure) for the Job/Pod that ran. If some post deploy step failed, I would like to see why. Does this make sense inline with the pipeline, or in the infrastructure tab? I don't know.
- This may be difficult in conjunction with the prior bullet, but automatic cleanup would be super nice too. Perhaps this could use the spinnaker max version history label to keep some around for logs but dispose the older ones?

**Comment by ethanfrogers** [comment]
@dchw We actually exposed logs in the V1 provider version of this stage. For better or for worse, we stored them in the pipeline execution context so they persisted after the Pod was cleaned up. This worked well for small jobs but caused issues for longer running tasks so we may consider a different solution in the future.

**Comment by ethanfrogers** [comment]
For anyone watching this issue:

i've been working on a POC of the above ^^ but wanted to get feedback from you all before I nail anything down. please take a look at this doc and let me know if you have any feedback. anyone should be able to comment on it. https://docs.google.com/document/d/1WDwJir_kQQij4REBsZ-70PvrWEG36yJdzeN0CUnW2Iw/edit?usp=sharing.

**Comment by ethanfrogers** [comment]
Update for anyone following - We've been working on this for the last few weeks and just landed the last little bit for the UI. Here are the details:
* There is now a `Run Job (Manifest)` stage available in the UI which accepts a Kubernetes Job text manifest (artifact support forthcoming). 
* The `Run Job (Manifest)` stage works the same way as the V1 version. Jobs are submitted and watched until completion and cleaned up. 
* With this comes support for Preconfigured Job stages which I'll be documenting. That should land in `spinnaker.io` by EOD tomorrow.

As for features, here are some things on my short list:
1. Consume job outputs into pipeline context.
2. Artifact support
3. Rethinking of how logs are captured. Currently stored in context, would like to eliminate that.

If you can think of any other features you'd like to see, let me know! Also, if you run into any issues while testing this please let me know!

**Comment by rverma-nikiai** [comment]
@ethanfrogers Is it released, I can't find run job[manifest] in 1.13.3

**Comment by ethanfrogers** [comment]
@rverma-nikiai it's been merged into `master` and will be released in 1.14.

**Comment by ethanfrogers** [comment]
The 1.14 release has been cut! For those of you who have been watching this issue you can test out the 1.14 pre-release with `hal config version edit --version release-1.14.x-latest-validated`! We've added some features that we think are going to make the Run Job stage a lot more useful and are excited to see how you use it!

Here's some resources to get you started - 
* [Documentation](https://spinnaker.io/guides/user/kubernetes-v2/run-job-manifest/)
* [Custom Job Stages](https://www.spinnaker.io/guides/operator/custom-job-stages/)
* [Blog post about Run Job](https://blog.spinnaker.io/extending-spinnaker-with-kubernetes-and-containers-5d16ec810d81)

I'm going to close this issue out for now. If you find any bugs or would like an improvement please file an issue!
